### PR TITLE
Add Catbox upload support

### PR DESCRIPTION
**Pull request by TheColorman**
This is a work in progress PR to add a button that uploads a video clip to [catbox.moe](https://catbox.moe) (free file hosting). So far I've only implemented Windows, but I thought I'd be a good idea to open up a PR for progress tracking and code review.  
  
I originally wanted some lua-only way of uploading the file, but I couldn't find anything that would work on all platforms with a reasonable bundle size, so I've opted to use platform-specific pre-installed tools such as cURL.  
  
<details>
  <summary>Screenshots</summary>
  
  
  
  
</details>

### CODE REVIEW

**Comment by TheColorman** [comment]
I'll do Linux next, but since I don't have a MacOS machine I won't be able to test on there. I can't imagine it's very different from Unix though.

**Comment by tatsumoto-ren** [comment]
All three operating systems have curl. On the GNU system it is available by default. So I imagine the code would be the same or similar regardless of the platform.

**Comment by TheColorman** [comment]
The clipboard code for Linux relies on either xclip or wl-copy, both of which (*I think*) are third party tools that need to be installed separately, so I want there to be a fallback, such as an output file, in case copying to clipboard isn't an option.
https://github.com/Ajatt-Tools/videoclip/blob/d49c40481688f05170ea7c47b3a4d17ed56eea8d/videoclip.lua#L497-L506

**Comment by TheColorman** [comment]
*Also*, these error handlers are built specifically for windows, I don't think the exit codes match the other systems.
https://github.com/Ajatt-Tools/videoclip/blob/d49c40481688f05170ea7c47b3a4d17ed56eea8d/videoclip.lua#L477-L484

File: videoclip.lua [hunk 1]
```diff
@@ -431,6 +448,69 @@ function main_menu:update()
...
+            -- This uses cURL to send a request to the cat-/litterbox API.
+            -- cURL is included on Windows 10 and up, most Linux distributions and macOS.
+
+            local r = mp.command_native({ -- This is technically blocking, but I don't think it has any real consequences ..?
+                name = 'subprocess',
+                playback_only = false,
+                capture_stdout = true,
+                capture_stderr = true,
+                args = {
+                    os_type == 'window' and 'curl.exe' or 'curl', '-s',
```
Comment (tatsumoto-ren) [inline videoclip.lua:467]: `os_type == 'window'` is never `true`. Calling `curl` should be sufficient on all systems. As far as I know, it shouldn't be necessary to append `.exe` to the executable name, even on Windows.

File: videoclip.lua [hunk 2]
```diff
@@ -217,6 +204,24 @@ local function notify(message, level, duration)
     mp.osd_message(message, duration)
 end
 
+------------------------------------------------------------
+-- System-dependent variables
+
+local os_type
```
Comment (TheColorman) [inline videoclip.lua]: Okay, turns out this is a problem because os_type was used at the top to determine the video folder for mac

Comment (TheColorman) [reply videoclip.lua]: b2dad49 is the best I got, maybe moving the utility functions to a separate file could make it cleaner? Thing for another day.

File: videoclip.lua [hunk 3]
```diff
@@ -431,6 +475,92 @@ function main_menu:update()
...
+                    }
+                    cb = subprocess(clipboard_command)
+                else
+                    -- wl-copy is from wl-clipboard
+                    local clipboard_command = {
+                        "wl-copy", r.stdout
+                    }
+                    cb = subprocess(clipboard_command)
+                end
+                if cb.status ~= 0 then
```
Comment (ayes-web) [inline videoclip.lua:551]: After the subprocess change the alternative action will always trigger since the command returns -3

File: videoclip.lua [hunk 4]
```diff
@@ -431,6 +475,92 @@ function main_menu:update()
...
+                        "/bin/sh", "-c", "'echo " .. r.stdout .. " | xclip -sel clip'"
+                    }
+                    cb = subprocess(clipboard_command)
+                else
+                    -- wl-copy is from wl-clipboard
+                    local clipboard_command = {
+                        "wl-copy", r.stdout
+                    }
+                    cb = subprocess(clipboard_command)
+                end
```
Comment (TheColorman) [inline videoclip.lua:550]: Using Wayland, it hangs here for no apparent reason, despite successfully copying.

**Comment by TheColorman** [comment]
I'm honestly not quite sure how to deal with the problems mentioned in https://github.com/Ajatt-Tools/videoclip/pull/20#discussion_r1234312358 and https://github.com/Ajatt-Tools/videoclip/pull/20#discussion_r1235672914. 
The hanging also used to happen to x11, but Ayes fixed this by wrapping it in `/bin/sh -c`.  
The exir code of `-3` indicates that `/bin/sh` could not find the command it is executing (e.g. `xclip`). Wrapping the Wayland command in `/bin/sh` also gives exit code -3 and fails to copy.  
  
To recap, both `xclip` and `wl-copy` hang after the copy (that is to say, the do succeed before hanging). Wrapping either in `/bin/sh -c "..."` causes the commands to not be found.

**Comment by TheColorman** [comment]
Fixed https://github.com/Ajatt-Tools/videoclip/pull/20#issuecomment-1599381497 by first checking if the clipboard binary is on path, and then copying in a detached process if found. I don't think this actually solves the underlying issue, so in theory the detached process still hangs after copying, but that only lasts until mpv is closed so it should be fine.

File: videoclip.lua (hunk 4 above)
Comment (TheColorman) [reply videoclip.lua]: See https://github.com/Ajatt-Tools/videoclip/pull/20#issuecomment-1600948684 / 77eb47c.

**Comment by ayes-web** [comment]
Imo the streamable button could be removed since this feature serves the intended (im assuming) purpose better

**Comment by ayes-web** [comment]
@tatsumoto-ren thoughts?

**Comment by TheColorman** [comment]
@tatsumoto-ren what's up? Anything I should change?

**Comment by tatsumoto-ren** [comment]
> thoughts?

Looks fine, but needs refactoring I guess.

File: videoclip.lua [hunk 5]
```diff
@@ -431,6 +475,90 @@ function main_menu:update()
     self:overlay_draw(osd:get_text())
 end
 
+function main_menu:upload_catbox()
+    local endpoint = config.litterbox and 'https://litterbox.catbox.moe/resources/internals/api.php' or 'https://catbox.moe/user/api.php'
+
+    encoder.create_clip('video',
+        function(outfile)
```
Comment (tatsumoto-ren) [inline videoclip.lua:482]: This function is too long for a "lambda", I think it is better to name it.

File: videoclip.lua [hunk 6]
```diff
@@ -431,6 +475,90 @@ function main_menu:update()
...
+                end
+            end
+
+            if os_type == "linux" then
+                local cb_x11 = string.find(subprocess({
+                    "whereis", "xclip"
+                }).stdout, "/")
+                local cb_wayland = string.find(subprocess({
+                    "whereis", "wl-copy"
+                }).stdout, "/")
```
Comment (tatsumoto-ren) [inline videoclip.lua:541]: You can have both xclip and wl-copy installed. Better if you do:
```
if is_wayland() then
clipboard_command = "wlcopy"
else
clipboard_command = "xclip"
end
```

File: videoclip.lua [hunk 7]
```diff
@@ -431,6 +475,90 @@ function main_menu:update()
...
+                    notify("Failed to copy URL to clipboard, trying to open in browser instead.\n(Make sure xclip or wl-clipboard is installed)", "warn", 6)
+                    mp.commandv('run', open_utility, r.stdout)
+                    return
+                end
+
+                if cb_x11 ~= nil then
+                    mp.command("run /bin/sh -c \"echo " .. r.stdout .. " | xclip -sel clip\"")
+                elseif cb_wayland ~= nil then
+                    mp.command("run wl-copy " .. r.stdout)
+                end
```
Comment (tatsumoto-ren) [inline videoclip.lua:553]: Just send the text to stdin to avoid problems with quotation marks.
``` lua
local clipboard_command = is_wayland() and "wlcopy" or "xclip -selection clipboard"
local handle = io.popen(clipboard_command, 'w') 
handle:write(text)
handle:close()
```

File: videoclip.lua [hunk 8]
```diff
@@ -431,6 +475,90 @@ function main_menu:update()
...
+                if cb.status ~= 0 then
+                    notify("Failed to copy URL to clipboard, trying to open in browser instead. (Make sure PowerShell is installed)", "warn", 4)
+                    mp.commandv('run', open_utility, r.stdout)
+                    return
+                end
+            end
+
+            if os_type == 'macos' then
+                local clipboard_command = {
+                    "pbcopy"
```
Comment (tatsumoto-ren) [inline videoclip.lua:525]: in mpvacious we use `LANG=en_US.UTF-8 pbcopy`.

File: videoclip.lua [hunk 9]
```diff
@@ -23,14 +23,37 @@ local mpopt = require('mp.options')
...
+    args = "uname",
+}).stdout
+
+if string.find(uname, "MSYS_NT") then
+    os_type = 'windows'
+elseif string.find(uname, "Darwin") then
+    os_type = 'macos'
+else
+    os_type = 'linux'
+end
```
Comment (tatsumoto-ren) [inline videoclip.lua:44]: We can have os_type as an enum (to avoid comparing strings directly).

File: videoclip.lua (hunk 6 above)
Comment (TheColorman) [reply videoclip.lua:541]: It was a lot cleaner originally, but I ran into a lot of problems trying to get the Linux commands to work (see https://github.com/Ajatt-Tools/videoclip/pull/20#issuecomment-1599381497), but tl;dr there's not way to run the command synchronously, and there's no way to get the exit code asynchronously to check if the command fails. I ended up opting for the current system that checks PATH for either of the binaries first, runs one if found, or opens in browser if not.

File: videoclip.lua (hunk 7 above)
Comment (TheColorman) [reply videoclip.lua]: The text shouldn't have any quotes, since it's a URL, but I guess it is technically untrusted so I'll change Linux and MacOS to this. I don't like using `io.popen` for Windows however, as it opens up a command window.

**Comment by tatsumoto-ren** [comment]
Refactored the code a bit. I've changed the way text is copied to the clipboard on GNU/Linux. It's the same implementation mpvacious uses. But feel free to revert it back if you notice that it doesn't work.

File: videoclip.lua (hunk 7 above)
Comment (tatsumoto-ren) [reply videoclip.lua]: Yes, there's no need to use `io.popen` on windows.

Comment (TheColorman) [reply videoclip.lua]: Another thing about `io.popen` is that there's no easy way to get exit codes without string splicing.

Comment (tatsumoto-ren) [reply videoclip.lua]: I guess calling `which` is still needed to test if wl-copy or xclip are installed.

Comment (TheColorman) [reply videoclip.lua]: Oh, didn't see you'd already refactored most of it when writing https://github.com/Ajatt-Tools/videoclip/pull/20#discussion_r1241283751, looks awesome.

Yeah I don't think there's a good way to test whether they're installed, and `which` is pretty messy. `io.popen` also returns a valid object even if the command it's running fails, so that's annoying.

Comment (TheColorman) [reply videoclip.lua]: Oh wow, I did some digging and turns out `io.popen` does in fact give you an exit code upon calling the `handle:close()` method. Specifically, it returns 3 values:
- `success`: `boolean`  
- `exitcode`: `"exit" | "signal"`  
- `code`: `integer`

Seems this'll work after all.

Comment (TheColorman) [reply videoclip.lua]: 6c571a8

File: videoclip.lua (hunk 6 above)
Comment (TheColorman) [reply videoclip.lua]: 6c571a8

File: videoclip.lua (hunk 8 above)
Comment (TheColorman) [reply videoclip.lua]: I'll try to see if passing environment variables works for subprocess, if not I'll switch it out for `io.popen` like the current Linux implementation.

File: platform.lua [hunk 10]
```diff
@@ -48,7 +48,14 @@ this.clipboard = (function()
...
         self.copy = function(text)
-            return h.subprocess({ self.clip_exe, }, text)
+            local handle = io.popen("LANG=en_US.UTF-8 pbcopy", 'w')
+            if handle then
+                handle:write(text)
+                local suc, exit, code = handle:close()
+                return { status = code }
+            else
+                return { status = 1 }
+            end
```
Comment (TheColorman) [inline platform.lua]: Don't entirely like how the exact same code is repeated twice for mac and linux now, maybe it makes sense to turn it into a function

Comment (tatsumoto-ren) [reply platform.lua]: Alright, merged the two conditions.
